### ISSUE TITLE
Models always over-fitting if left training.

### ISSUE THREAD

**Issue by dgriffiths3** [fizyr/keras-retinanet#460]
I've trained a few networks using this repository and get pretty good results. However, I have one observation that I wanted to ask about. All of my models seem to start over-fitting if left training. The accuracy never seems to just converge. This usually takes the form of an initial high increase in mAP accuracy (whilst loss is still high) followed by a slow decline. I'm just a little confused how the highest accuracy occurs when the loss is still relatively high.

I was thinking this could be caused by a) an issue with the implementation, b) conflict between my training data and retinanet, c) A common trait of retinanet or d) something else?

[Here's a link](https://www.dropbox.com/s/rlaaft1dqov4wra/overfit.png?dl=0) to two examples.

**Comment by hgaiser** [comment]
I haven't observed this when training COCO at least, haven't monitored other datasets closely enough to say if it occurs or not. How big is your dataset? You could try to do random transform augmentation to increase your dataset.

**Comment by dgriffiths3** [comment]
@hgaiser My data set in the images is ~25,000 images so should be big enough. It converges well on a Faster R-CNN model so not sure why it won't here. I've also had the same issue when training on a smaller data set (~1000 images) with and without data augmentation.

**Comment by hgaiser** [comment]
That should be big enough, I agree. Strange.. I wouldn't know why there's such a big case of overfitting in your dataset. If I think of something I'll let you know. If you find out what caused this (or what fixed it maybe), I'd be interested to hear about it.

**Comment by JudyMRSD** [comment]
If I add an early stopping to training, what would be a good value to check to decide wether to stop? If I stop when average iou for evaluation set is not increasing, would this be a good criteria? Another option I was thinking about is use loss for early stopping. I'm not sure if I can use loss value for early stopping , since I'm not sure if it's better to use regression loss and classification loss. Thanks a lot!

**Comment by hgaiser** [comment]
> If I add an early stopping to training, what would be a good value to check to decide wether to stop? If I stop when average iou for evaluation set is not increasing, would this be a good criteria? Another option I was thinking about is use loss for early stopping. I'm not sure if I can use loss value for early stopping , since I'm not sure if it's better to use regression loss and classification loss. Thanks a lot!

You can use total loss as a metric too. The thing is, you can create an output that has a terrible loss value, but gets perfect testing score. So checking for loss is better than nothing, but it doesn't guarantee a good result.

**Comment by apacha** [comment]
As Geoffrey Hinton put it: "Early stopping (is) beautiful free lunch". I recommend the following approach: 
- Check out the changes that were introduced by this [PR](https://github.com/fizyr/keras-retinanet/pull/468)
- Make sure, you are monitoring your training performance on an independent validation set and the validation metric (e.g. mAP) gets written to the logs
- Attach an [EarlyStopping](https://keras.io/callbacks/#earlystopping) callback and monitor that metric, same as for the ModelCheckpoint callback [here](https://github.com/fizyr/keras-retinanet/pull/468/commits/463872c3987693a0e752ecad90e586a82a666d7e#diff-fb5be6ca95a3190ebad5eaedf0a5ce65R145). Stop if the mAP on the validation set does not improve for a few epochs, e.g. 8.

Ps: Please attach images directly to your post, because your dropbox link is already dead.

**Comment by JudyMRSD** [comment]
Thank you very much! I will try the newly added early stopping!

**Comment by apacha** [comment]
Well, we haven't added it yet. But we prepared something, to make it easy to add it. 
@hgaiser what do you say on actually adding it? It must take care of the right metric and should only be added, if there is a validation happening during training, but I see it as a useful addition and Keras is very convenient for that behavior.

**Comment by hgaiser** [comment]
I would be in favour of adding it. I think the default behaviour should stay the same though. We'd need to add a `"mAP"` log for COCO though, otherwise it won't work there.

**Comment by wajahat57** [comment]
Is you training data different than test data? Such as synthetically generated training data but real test data.

**Comment by hgaiser** [comment]
I'm closing this issue, since I believe it's not an issue with the code, but more likely an "issue" with the algorithm itself.
